### ISSUE TITLE
Use with IRSA

### ISSUE THREAD

**Issue by shaunc** [s3fs-fuse/s3fs-fuse#1778]
### Additional Information
_The following information is very important in order to help us to help you. Omission of the following details may delay your support request or receive no attention at all._
_Keep in mind that the commands we provide to retrieve information are oriented to GNU/Linux Distributions, so you could need to use others if you use s3fs on macOS or BSD_

#### Version of s3fs being used (s3fs --version)
1.85

#### Version of fuse being used (pkg-config --modversion fuse, rpm -qi fuse, dpkg -s fuse)
2.9.9

#### Kernel information (uname -r)
4.14.186-146.268.amzn2.x86_64

#### GNU/Linux Distribution, if applicable (cat /etc/os-release)
NAME="Ubuntu"
VERSION="20.04.3 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.3 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

### Details about issue

I am looking for instructions for using with IRSA. In my pod container environment I have AWS_ROLE_ARN and AWS_WEB_IDENTITY_TOKEN_FILE. How can I get s3fs-fuse to consume these?

**Comment by gaul** [comment]
Sorry I'm not familiar with with IRSA.  Later versions of s3fs do support IMDSv2 -- does this work for you?

@nmeyerhans do you know anything about this?

**Comment by theis188** [comment]
Thanks for opening this issue, I would also find this feature useful.

**Comment by aru-trackunit** [comment]
+1

**Comment by bressanmarcos** [comment]
:+1: +1

**Comment by mak-454** [comment]
+1

**Comment by yao531441** [comment]
+1

**Comment by yukinakanaka** [comment]
+1

**Comment by ggtakec** [comment]
As one solution to this issue:

I have released the following source as a prototype few month ago.
https://github.com/ggtakec/s3fs-fuse-awscred-lib
_You can refer to the README.md for how to build/use it._
This library is a DSO that wraps the `aws-sdk-cpp` authentication functionality.
The code in the master branch of s3fs-fuse includes an option(`credlib`) to delegate authentication to an external DSO library.
By specifying this library, I think that functions such as authentication tokens provided by AWS works.

If anyone has tried it, I'd love to know the results.
Could anyone use the code from the master branch of s3fs-fuse to try out the `libs3fsawscred.so` file that the above repository creates?

**Comment by artificial-aidan** [comment]
I tried it and ended up needing to make these changes for it to work: https://github.com/artificialinc/s3fs-fuse/tree/aidan/ext-creds

It was failing early for me, and then the session token wasn't being set.

I'd suggest not logging the access keys either. I know you have to turn trace on, but seems dangerous.

**Comment by ggtakec** [comment]
@artificial-aidan Thanks for your kindness.
As you pointed out, it was a buggy code.
I am creating a PR #2101 based on your code.(If merged, it will be reflected in the master code.)

If possible, please let me know about `It was failing early` that you tried in your environment.
I would like to know the log when the `dbglevel=info`(and `curldbg`: but this option makes log file large) option is given when starting s3fs.
Also, although the log becomes complicated, if you add `credlib_opts=Info` along with `credlib=libs3fsawscred.so`, you can also record the trace log of `libs3fsawscred.so`.

**Comment by artificial-aidan** [comment]
Yeah I should be able to reproduce it. I will look back at my logs on Monday.

**Comment by artificial-aidan** [comment]
Reverting back to the old code was more work than I had time for.

I ran into 2 issues. The first was the error `could not determine how to establish security credentials.` This is because when using IRSA, like ECS, we need to delay the loading of the credentials. This might cause issues if you were using the external library to load other credentials, I'm not sure. But for IRSA, it needs to be delayed

The second issue was the error message was when I actually got the creds, but then s3fs tried to make a call, and I got a permission denied error. This error turned out to be from not setting the session token.

Here is the output of a successful run:

```
FUSE library version: 2.9.9
nullpath_ok: 0
nopath: 0
utime_omit_ok: 1
unique: 2, opcode: INIT (26), nodeid: 0, insize: 56, pid: 0
INIT: 7.31
flags=0x03fffffb
max_readahead=0x00020000
[s3fsawscred] : Access Key Id = XXXXXXXX
[s3fsawscred] : Secret Key    = XXXXXXXX
[s3fsawscred] : Session Token = XXXXXXXXX
[s3fsawscred] : expiration    = 2023-01-30TXXXXXX
   INIT: 7.19
   flags=0x00000011
   max_readahead=0x00020000
   max_write=0x00020000
   max_background=0
   congestion_threshold=0
   unique: 2, success, outsize: 40
```

**Comment by ggtakec** [comment]
@artificial-aidan Sorry about the late reply.

> Reverting back to the old code was more work than I had time for.

Thank you for your great help.

Since there was a correction omission earlier, I corrected #2101 a little.

And I would like to reconfirm with you whether this problem still remains or not.
Did #2101(or the modified code you originally tried) solve this problems you had?
Or are you still getting `session token not set`, etc?

I would like to merge #2101 if it solves this problem.
If this problem still remains, I will prepare the environment to use IRSA and investigate, so it will take some time.

Thanks in advance for your assistance.

**Comment by artificial-aidan** [comment]
Let me try it out Monday. 

The branch I shared was fully functional for me.

**Comment by ggtakec** [comment]
@artificial-aidan 
#2101 has been merged.
Please check and let me know if you have any problems.

**Comment by bmihaescu** [comment]
@ggtakec I also have enabled IRSA on an EKS pod and it seems that it's still trying to get the role/credentials from instance metadata. What do I need to change to the s3fs command to use the changes you merged with #2101?

I use V1.92 version of s3fs 

Thanks

**Comment by bmihaescu** [comment]
> https://github.com/artificialinc/s3fs-fuse/tree/aidan/ext-creds

I want to mention that on EKS we have env vars as part of IRSA: 
AWS_ROLE_ARN=
AWS_WEB_IDENTITY_TOKEN_FILE=

**Comment by ggtakec** [comment]
@bmihaescu 
Currently, s3fs does not have IRSA authentication functionality.
Options(`credlib` and `credlib_ops`) added in #1927(#2101)  are the ability to load external authentication libraries into s3fs.
Currently, it seems that the only library that supports this is [s3fs-fuse-awscred-lib](https://github.com/ggtakec/s3fs-fuse-awscred-lib).
For how to use this [s3fs-fuse-awscred-lib](https://github.com/ggtakec/s3fs-fuse-awscred-lib) library refer to its README.
This library uses the `aws-sdk-app` and supports all authentication methods provided by AWS(except SSO).
If you have any questions about this library, please open as an issue on [s3fs-fuse-awscred-lib](https://github.com/ggtakec/s3fs-fuse-awscred-lib) github.
Please let me know if I misunderstood your question.
